### ISSUE TITLE
MMOD loss goes negative when target objects overlap more than allowed by NMS

### ISSUE THREAD

**Issue by reunanen** [davisking/dlib#870]
For some reason it happens on one specific dataset that the MMOD loss goes negative. I believe this should not be possible?

It seems to me that the problem has to do with the positive confidence going higher than it should (perhaps because some of the samples are very easy to learn, or something like that).

More specifically, it seems that the loss goes negative because [here](https://github.com/davisking/dlib/blob/c2a9dee846cad7b20998ee0f3fae79b6bcb67d43/dlib/dnn/loss.h#L765) some of the `out_data` are large – much larger than the `loss_per_missed_target` that [is added](https://github.com/davisking/dlib/blob/c2a9dee846cad7b20998ee0f3fae79b6bcb67d43/dlib/dnn/loss.h#L751) to the loss initially (fwiw, I'm using the default losses of 1.0 for both missed targets and false alarms).

I have implemented [a quick fix](https://github.com/reunanen/dlib/commit/e33959968152014ac5b0f9db087cd7581b8f5f6d) that seems to help at least in my case: the loss no longer goes negative, and also the end results are more reasonable and more stable.

However, I do not think my fix is overall correct. In particular the conditional gradient update appears inelegant, if not outright wrong. Nevertheless, in my case the change very clearly helps keep the loss and the gradients sane.

**Comment by davisking** [comment]
Have you trained models with this change?  I would expect it to mess up the
results.

Anyway, the reason this happens is because you have target objects that are
impossible to detect, probably due to how NMS is configured.  You should
get warnings about this from the code though.  You don't get any warning
messages?

**Comment by reunanen** [comment]
Thanks again for the quick and helpful response!

> Have you trained models with this change?  I would expect it to mess up the
> results.

Yes, I have – but only with "my" data, not your face or vehicle sets.

I too expected this change (which was not much more than an uneducated guess) to mess up the results... but to my surprise training seems to work all right, at least for a while and when compared to the version from before the change.

> Anyway, the reason this happens is because you have target objects that are
> impossible to detect, probably due to how NMS is configured.  You should
> get warnings about this from the code though.  You don't get any warning
> messages?

Yes, I do get warnings. (Or in fact I don't, because I [added an option to hide these warnings](https://github.com/reunanen/dlib/commit/067dc535b8e7a7441ef93dcde96a7e9b41c6ef03).)

That's because I deliberately set [the NMS thresholds](https://github.com/davisking/dlib/blob/c2a9dee846cad7b20998ee0f3fae79b6bcb67d43/dlib/image_processing/box_overlap_testing.h#L117-L118) to be lower (= allow less overlapping results) than what would be calculated from the training data. I guess that's a bad idea, then.

My training data contains some near-dupes that overlap a lot. It's not exactly incorrect considering the application domain, however if I just use the `mmod_options` constructor, I get very high thresholds (=allow lots of overlap), which just makes inference results look next to crazy. I thought I can simply decrease the thresholds, and indeed it looked like it worked quite well, so I even sort-of forgot I had done this. Or at least didn't think about it in this context.

So. The correct fix is not to decrease the thresholds, but rather decimate the training samples so that it's at least possible to detect all the target objects... Right?

**Comment by davisking** [comment]
Never disable those warnings.  A while back I had the code set to terminate execution if they happened.  But that was a little bit too harsh since it's possible to create training data that only triggers the warning once every few hours or at some similarly low rate that is not consequential.  In those cases, aborting execution is frustrating to the user and not really necessary.  

Anyway, those warnings are there for a reason.  You can't give the trainer examples of objects it should detect that are impossible for it to detect.  That's why the loss goes negative.  There is also an extended discussion of this here: http://dlib.net/dnn_mmod_train_find_cars_ex.cpp.html.  The short story is that you need to fix your training data.

**Comment by reunanen** [comment]
Ok, got it. Thanks again for the help!

**Comment by davisking** [comment]
No problem :)
